Force exits in PushSubscription PUT, POST and DELETE handlers

MrPetovan opened the pull request:
Fix #14325
Fix #14450 
Fix #14469
Reverts #14326
Reverts #14484
Reverts #14530

I decided to revisit #14326 which seemed too extensive to me given the localized issue it was trying to fix. In particular, it was meant to address `rawContent()` but it affected `content()` as well which we expected to be able to be called after a successful call to `post()`, which prompted the subsequent #14484 and #14530 quick fixes.

Instead, I fixed the original issue reported in #14325 which concerns only a couple modules. If PUT, PATCH, POST and DELETE handlers set their own data, then they must exit themselves through `BaseModule->httpExit()` or `BaseModule->jsonExit()` in this case.

An earlier version of this PR tried to fiddle with the `BaseModule->run` order of operations, but it tripped the `Options` module that should execute `rawContent` in a non-GET context. Thanks to @nupplaphil for providing a crucial unit test to avoid a collateral issue with my naive fix.

Also thanks to @bmillwood for digging up the cause of the issue they reported, it proved crucial to provide a simple solution to the UX problem caused by #14326

MrPetovan committed:
Revert "QuickFix: 2FA - App specific password"

This reverts commit dc4eac2742c09199e0a8a6a85c0129a7097e4a9d.

MrPetovan committed:
Update messages.po main translation file

MrPetovan committed:
Revert "Fix installation"

This reverts commit 67a8c5073cc45c1f9beb8f25d66f1d64540950de.

MrPetovan committed:
Revert "Issue 14325: Process "post" requests only once"

This reverts commit 0b710ce4fb7443ecdac83a058c0de0cf4dfe8772.

MrPetovan committed:
Force exits in PushSubscription PUT, POST and DELETE endpoints

- These add their own content to the response, which makes rawContent() redundant

MrPetovan commented:
@nupplaphil The Code Standards Check fails, but running the bespoke fix script didn't work:
```
$ bin/dev/fix-codestyle.sh
PHP CS Fixer 3.62.0 7th Gear by Fabien Potencier, Dariusz Ruminski and contributors.
PHP runtime: 8.3.0
Running analysis on 1 core sequentially.
You can enable parallel runner and speed up the analysis! Please see usage docs for more information.
Loaded config default from ".php-cs-fixer.dist.php".
    0 [░░░░░░░░░░░░░░░░░░░░░░░░░░░░]
Found 0 of 0 files that can be fixed in 0.023 seconds, 14.00 MB memory used
```

annando commented:
I don't really like, that `rawContent` and `content` are now executed all the time. I think that we have to have a dedicated function only for `GET`, especially for API calls.

bmillwood commented:
> I don't really like, that `rawContent` and `content` are now executed all the time. I think that we have to have a dedicated function only for `GET`, especially for API calls.

Choosing to have a dedicated function seems fine, but:
- you need *some* means to display content on POST pages, including content that is in some cases almost the same as the GET content,
- you also need to figure out how to identify where in the existing code assumes that `content()` is called on POST requests, besides the two cases we've already found (is it possible that addons could be relying on that behaviour?)

bmillwood commented:
It does seem potentially worth treating the cases of API requests vs. ordinary web UI usage differently, at least conceptually.

annando commented:
@bmillwood one solution could be to implement a dedicated `get` function. The problem really lies in the API functions that allow multiple actions depending on the used method.

MrPetovan commented:
Here are the conflicting requirements that led to this solution:
1. We have dedicated module handlers for PUT, POST, PATCH, DELETE but not for OPTIONS.
2. We expect some POST requests to also run `content()` afterwards including information set in `post()`.
3. We expect some PUT, POST, PATCH and DELETE requests to also return content but not run `rawContent()` afterwards.
4. We have two different handlers for GET.
5. We have a catch-all handler including for OPTIONS.
6. We can't know if the module selected by the router implements `rawContent()`.

We could have a system where `rawContent()` isn't always called if we removed requirements 5 and 6.

Basically:
- Add a dedicated handler for OPTIONS. It would likely be set at the `BaseModule` level instead of being a special module on its own.
- Turn `rawContent()` into a GET-only handler.
- Only execute it if the module implements it through a dedicated interface.
- Return early if `rawContent()` was executed.

This is very extensive though, do you think it's worth not having `rawContent()` potentially executed all the time?

annando commented:
Have a look for example for the routes of `/statuses/{id:\d+}`, `/push/subscription` or `/lists/{id:\d+}`. We shouldn't - for example - try to fetch a given list after just deleted them. Same is true for calling `PUT` on a status.

A simple solution is to introduce a `get` function that we can execute in all places where we really shouldn't executed `rawContent`.

bmillwood commented:
Another (minor) thing to mention is that the docs for `post()` in `BaseModule.php` say `Doesn't display any content`. Presumably this wouldn't be true of `get()`, and I suspect it may already not be true of `post()` in some cases. So you might want to think about how to amend that doc.

MrPetovan commented:
> Have a look for example for the routes of `/statuses/{id:\d+}`, `/push/subscription` or `/lists/{id:\d+}`. We shouldn't - for example - try to fetch a given list after just deleted them. Same is true for calling `PUT` on a status.
> 
> A simple solution is to introduce a `get` function that we can execute in all places where we really shouldn't executed `rawContent`.

It isn't simple for the reasons I listed earlier. Even if we just added a new `get()` method we would still need to call `rawContent` at all times. And the changes would be extensive as well.

MrPetovan commented:
> Another (minor) thing to mention is that the docs for `post()` in `BaseModule.php` say `Doesn't display any content`. Presumably this wouldn't be true of `get()`, and I suspect it may already not be true of `post()` in some cases. So you might want to think about how to amend that doc.

I wrote this without taking specific API endpoints in consideration that return the equivalent of `get()` for the entity that is created/updated/deleted. It doesn't make much sense to run `get()` on the same entity a second time, especially for deletion, so I like that we add meaningful content to the response. However it is raw JSON, not a templated page.

nupplaphil committed:
Make PHP-CS happy

nupplaphil commented:
> @nupplaphil The Code Standards Check fails, but running the bespoke fix script didn't work:
> 
> ```
> $ bin/dev/fix-codestyle.sh
> PHP CS Fixer 3.62.0 7th Gear by Fabien Potencier, Dariusz Ruminski and contributors.
> PHP runtime: 8.3.0
> Running analysis on 1 core sequentially.
> You can enable parallel runner and speed up the analysis! Please see usage docs for more information.
> Loaded config default from ".php-cs-fixer.dist.php".
>     0 [░░░░░░░░░░░░░░░░░░░░░░░░░░░░]
> Found 0 of 0 files that can be fixed in 0.023 seconds, 14.00 MB memory used
> ```

`COMMAND=fix TARGET_BRANCH=develop bin/dev/fix-codestyle.sh`

Currently it fails because of the un-rebased changes

annando commented:
@MrPetovan it is no problem to execute `rawContent` once we have `get`. Because then we can move all API functions from `rawContent` to `get`, so that there won't be any call to `rawContent` for these API calls any more.

MrPetovan commented:
> > @nupplaphil The Code Standards Check fails, but running the bespoke fix script didn't work:
> 
> `COMMAND=fix TARGET_BRANCH=develop bin/dev/fix-codestyle.sh`
> 
> Currently it fails because of the un-rebased changes

I think I understand what the command does but at the time it was correctly rebased. Thanks for fixing this branch!

MrPetovan commented:
> @MrPetovan it is no problem to execute `rawContent` once we have `get`. Because then we can move all API functions from `rawContent` to `get`, so that there won't be any call to `rawContent` for these API calls any more.

> move all API functions from `rawContent` to `get`

That's what I mean by "extensive changes". It's a simple change but in hundreds of files. This PR is simpler and address pressing UX concerns. Please feel free to create the `get()` method in a separate PR.

MrPetovan commented:
I nuked the tests 😩

nupplaphil commented:
> I nuked the tests 😩

something really strange happened ... git.friendi.ca temporary returned 502 for friendica-addons

annando commented:
See PR https://github.com/friendica/friendica/pull/14675

MrPetovan commented:
Thanks but I meant your PR as a follow-up to this one which first need to be merged.

annando commented:
My PR is a preparation for your PR to avoid API problems that would occur if your PR was in the system before.

MrPetovan commented:
I've merged your PR first and rebased mine onto yours. None of the changes I added were rendered unnecessary as a result, I'm a little puzzled. In particular `rawContent()` can still be executed after `get()`.
